ydb/public/lib/ydb_cli/import/import.cpp

```diff
@@ -585,49 +691,61 @@ TStatus TImportFileClient::UpsertParquet([[maybe_unused]] const TString& filenam
             break;
         }
 
-        const TString strSchema = NYdb_cli::NArrow::SerializeSchema(*batch->schema());
-        const size_t totalSize = NYdb_cli::NArrow::GetBatchDataSize(batch);
-        const size_t sliceCount =
-            (totalSize / (size_t)settings.BytesPerRequest_) + (totalSize % settings.BytesPerRequest_ != 0 ? 1 : 0);
-        const i64 rowsInSlice = batch->num_rows() / sliceCount;
+        auto upsertParquetBatch = [&, batch = std::move(batch)]() {
+            const TString strSchema = NYdb_cli::NArrow::SerializeSchema(*batch->schema());
+            const size_t totalSize = NYdb_cli::NArrow::GetBatchDataSize(batch);
+            const size_t sliceCount =
+                (totalSize / (size_t)settings.BytesPerRequest_) + (totalSize % settings.BytesPerRequest_ != 0 ? 1 : 0);
+            const i64 rowsInSlice = batch->num_rows() / sliceCount;
 
-        for (i64 currentRow = 0; currentRow < batch->num_rows(); currentRow += rowsInSlice) {
-            std::stack<std::shared_ptr<arrow::RecordBatch>> rowsToSend;
+            for (i64 currentRow = 0; currentRow < batch->num_rows(); currentRow += rowsInSlice) {
+                std::stack<std::shared_ptr<arrow::RecordBatch>> rowsToSendBatches;
 
-            if (currentRow + rowsInSlice < batch->num_rows()) {
-                rowsToSend.push(batch->Slice(currentRow, rowsInSlice));
-            } else {
-                rowsToSend.push(batch->Slice(currentRow));
-            }
+                if (currentRow + rowsInSlice < batch->num_rows()) {
+                    rowsToSendBatches.push(batch->Slice(currentRow, rowsInSlice));
+                } else {
+                    rowsToSendBatches.push(batch->Slice(currentRow));
+                }
 
-            do {
-                const auto rows = rowsToSend.top();
+                do {
+                    auto rowsBatch = std::move(rowsToSendBatches.top());
+                    rowsToSendBatches.pop();
 
-                rowsToSend.pop();
-                // Nothing to send. Continue.
-                if (rows->num_rows() == 0) {
-                    continue;
-                }
-                // Logarithmic approach to find number of rows fit into the byte limit.
-                if (rows->num_rows() == 1 || NYdb_cli::NArrow::GetBatchDataSize(rows) < settings.BytesPerRequest_) {
-                    // Single row or fits into the byte limit.
-                    auto status = WaitForQueue(inFlightGetter.GetCurrentMaxInflight(), inFlightRequests);
-                    if (!status.IsSuccess()) {
-                        return status;
+                    // Nothing to send. Continue.
+                    if (rowsBatch->num_rows() == 0) {
+                        continue;
                     }
+                    // Logarithmic approach to find number of rows fit into the byte limit.
+                    if (rowsBatch->num_rows() == 1 || NYdb_cli::NArrow::GetBatchDataSize(rowsBatch) < settings.BytesPerRequest_) {
```

Comment:
It's better to make SerializeBatchNoCompression() and check result data size over settings.BytesPerRequest_. Then GetBatchDataSize() is not needed.

Reply:
If we serialize batch using `SerializeBatchNoCompression` and result size will be `>= settings.BytesPerRequest_`, then we must somehow split batch, and throw away serialized result. 
Method `GetBatchDataSize` is not heavy, so it does not lead to some performance issues.